### PR TITLE
Cache asset compilation results on CI

### PR DESCRIPTION
**Pull request by mjankowski**
The `test-ruby` CI workflow hsa an initial `build` job which all of the other test jobs rely on, and therefore which must run first before any of them start. Within this job there's a matrix that has separate production and test modes.

The jobs are setting up ruby envs, JS envs, and then running the asset compilation step. The production mode is just a smoke test to make sure it runs successfully. The test mode uses artifact upload to make the compiled assets available to the subsequent jobs, which download them at their start.

However, within the job itself, there's no caching from run-to-run, which means that we have a ~1+ minute compilation step (the rails propshaft compile is part of this, but is pretty fast; the webpack compile step is the bulk of it) which runs every single time even if nothing assets-related has changed. I've heard that upcoming vite changes make this step a lot faster, but it's not clear to me how imminent that is.

This change adds a cache to the compilation step which uses the webpack native cache check (uses a digest of files in the manifest) so that if there's not actually any changes that need a new compile, it will just re-use whatever it already has. This is similar to how the stock local testing webpack config would work. On my fork this drops the step from ~1:20 or so down to < 10s on cache hits.

Configuration based in part on a sample config in the webpack docs and some related configs I saw in the wild. Would love a review here of the cache keys and dirs included to make sure this will work like I think it does.

Possible future work here - remove the artifact upload step and just rely on the cache entirely? This may not matter much - in the grand scheme of things that is a tiny fraction of the time in the run - but it does seem weird to move the same stuff two different ways. I wanted to keep this PR as just the speed-up of the compilation step.

### CODE REVIEW

**Comment by ClearlyClaire** [comment]
Beware that webpacker has really strange failure modes when webpack fails, and half of the time will claim that everything was built correctly on successive runs.

Though I guess in this case the failure would cause the task itself to fail, and the cache to not be persisted?

**Comment by mjankowski** [comment]
Yes I assume a failure of that sort would be obvious and then subsequent pushes to resolve would have new sha/ref that would make new cache on successful runs.

**Comment by mjankowski** [comment]
Rebased this to see cache work or not, seems like it did:

- Initial cache miss run - https://github.com/mastodon/mastodon/actions/runs/9762686445/job/26946742177 - took ~1:20 to compile assets
- Cache hit run - https://github.com/mastodon/mastodon/actions/runs/9762907872/job/26947468101 - restored from cache and took ~8s to compile (I assume this is mostly the propshaft compile and some small part is webpack doing its cache check)

**Review by renchap — APPROVED** [review]

**Comment by oneiros** [comment]
I understand much too little of all this to make a smart comment, but one thing I wonder:

> The test mode uses artifact upload to make the compiled assets available to the subsequent jobs, which download them at their start.

Is this still needed with the caching in place?

**Comment by mjankowski** [comment]
I had noted that at the very very end of my (long) PR description here ... basically, I think we can probably further optimize on that front, but am generally just trying to do one thing at a time in these optimization PRs.

**Comment by oneiros** [comment]
> I had noted that at the very very end of my (long) PR description here ... basically, I think we can probably further optimize on that front, but am generally just trying to do one thing at a time in these optimization PRs.

It is definitely too hot and humid today for my brain to work properly. I am sorry for the noise. :man_facepalming:

**Review by oneiros — APPROVED** [review]

**Comment by mjankowski** [comment]
Curious on status of this one -- we have two approvals but no merge yet. Should save roughly ~1 min on every CI run.

My personal opinion as the author is that this is pretty low risk ... I'm sure it's **possible** that there's an edge case here with webpacker that this will mess up, but we will likely see that and can revert if so.

If/when this does merge, I'll look into the artifacts-vs-caching thing as followup.

**Comment by renchap** [comment]
Woops, I thought I merged it. Let's see how it behaves and evaluate the results in 1-2 weeks, and if this created unforeseen issues?

**Comment by mjankowski** [comment]
Yup, makes sense to monitor ... will hold off on further improvement until we're reasonably sure that it's both a) faster, b) not broken.

**Pull request merged**
